Add large_types_passed_by_value lint

cauebs opened the pull request:
Creates a new lint that checks for large function parameters passed by value. ~Types that are Copy were ignored, because I understand they are explicitly marked as being cheap (or just acceptable) to copy. Arrays were excluded from that restriction because they are always Copy, independently of the size, if the elements are Copy.~ Only Copy types are considered, because if a type is not Copy it cannot be dereferenced, as pointed out by @ebroto, and that would require analyzing the whole function body to check if the argument is moved somewhere else. `self` and `mut` parameters are also skipped.

I refactored `trivially_copy_pass_by_ref` and the new lint into a new `pass_by_ref_or_value` module, since both share most of their implementations, as was pointed out to me in #4499.

~Some questions that remain:~
~1. Should `self` be disregarded for this lint?~
~2. Should we special case types like Vec and String when suggesting changes? (to slices and &str)~
~3. What should be the default size limit?~
~4. Are there any special cases I'm missing?~

I ask the maintainers to add the `hacktoberfest-accepted` label if this PR is decent, even if there are some minor details to work out, but I do intend to stick around and finish the job.

Fixes #4499

changelog: Add new lint [`large_types_passed_by_value`]

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @matthiaskrgr (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md) for more information.

ebroto reviewed (changes requested):
Thanks for your contribution to Clippy! I've left some remarks.

About your open questions:
1) I think we should exclude `self` as you suggest because it's a well-known pattern to consume a value, regardless of the efficiency.
2) `Vec` and `String` should never trigger the lint as they are quite small in the stack (and are not `Copy`, see my other comment).
3) and
4) I hope I've answered those in the review itself.

Also, the empty stdout file should be removed.

---

cc @flip1995 if you have the time/energy, I would not mind another pair of eyes on this, as some of my suggestions are quite opinionated :)

ebroto commented:
Also, FYI we should fix the dogfood test. Hopefully after the suggestions are applied there should be less (if anything) to modify.

flip1995 commented:
> I ask the maintainers to add the hacktoberfest-accepted label if this PR is decent, even if there are some minor details to work out, but I do intend to stick around and finish the job.

That shouldn't be necessary. PRs are automatically accepted in this repo:

```
PRs count if:
Submitted during the month of October AND 
Submitted in a public repo AND (
  The PR is labelled as hacktoberfest-accepted by a maintainer OR
  Submitted in a repo with the hacktoberfest topic AND (            <-- This applies for Clippy
    The PR is merged OR
    The PR has been approved
  )
)
```

flip1995 commented on clippy_lints/src/pass_by_ref_or_value.rs:
```diff
@@ -54,14 +54,41 @@ declare_clippy_lint! {
+    ///
+    /// **Why is this bad?** Arguments passed by value might require a shallow copy,
+    /// which can be expensive and can be easily improved with passing by reference.
+    ///
```
Shouldn't we rather just prevent this lint from triggering on `pub` functions?

ebroto replied:
I thought about it but since the other lint was not doing it, I thought it was preferable to stick with the same behavior.

flip1995 replied:
I think it is fine for the other lint, because of auto-deref. But there is no auto-ref, so I would prevent this lint on pub functions.

cauebs replied:
I'm hesitant both because this would require a lot of changes in the implementation (`Fn`s don't have `vis`; `Item`s, `ImplItem`s and `TraitItem`s do), and because I think the developers should be able to decide whether to make that breaking change or not. The function might be marked public from the very beginning and this lint won't show up to help. The responsability of maintaining API compatibility is on the developers, and I think it's fine if we just warn in the documentation.

ebroto replied:
>I think the developers should be able to decide whether to make that breaking change or not
>The responsability of maintaining API compatibility is on the developers

Of course, but I agree with @flip1995 that it's better to not encourage breaking changes if we can avoid it. Some folks will not read the lint documentation and follow the suggestion without taking into consideration the implications.

>this would require a lot of changes in the implementation (Fns don't have vis; Items, ImplItems and TraitItems do)

You should be able to use `cx.access_levels.is_exported(hir_id)`

flip1995 commented on clippy_lints/src/pass_by_ref_or_value.rs:
```diff
@@ -96,37 +127,55 @@ impl<'tcx> TriviallyCopyPassByRef {
+                }
+            } else {
+                if_chain! {
+                    if !is_copy(cx, ty) || matches!(ty.kind(), ty::Array(_, _));
```
+1 on all of this. I don't think it makes sense for non-`Copy` types, since they are moved and not copied anyway. That means the compiler doesn't need to allocate new space for the moved value.

cauebs replied:
Hum... Aren't values likely copied on move regardless of being `Copy`? I understand you have to "mov" the contents from the stack to registers to do a function call, but I might be wrong. Genuinely curious.

ebroto replied:
Well, I think in the general case space on the stack will be reserved for function parameters, and the value passed as an argument (moved) will be copied into that space.

flip1995 commented on clippy_lints/src/pass_by_ref_or_value.rs:
```diff
@@ -96,37 +127,55 @@ impl<'tcx> TriviallyCopyPassByRef {
+                            &format!("this argument ({} byte) is passed by value, but would be more efficient if passed by reference (limit: {} byte)", size, self.value_max_size),
+                            "consider passing by reference instead",
+                            format!("&{}", snippet(cx, input.span, "_")).into(),
+                            Applicability::MaybeIncorrect,
```
Rather `MaybeIncorrect`? I mean the suggestion is correct locally, but may require refactoring afterwards.

`Unspecified` is when we are not sure if the suggestion is even applicable (or someone was too lazy to find out what the applicability actually is (maybe that someone was me back in the day) 😄 )

ebroto replied:
Oh my bad, I think I was trying to be too conservative to avoid build errors.

flip1995 commented on clippy_lints/src/utils/conf.rs:
```diff
@@ -150,6 +150,8 @@ define_Conf! {
     /// Lint: TRIVIALLY_COPY_PASS_BY_REF. The maximum size (in bytes) to consider a `Copy` type for passing by value instead of by reference.
     (trivial_copy_size_limit, "trivial_copy_size_limit": Option<u64>, None),
+    /// Lint: LARGE_TYPE_PASS_BY_MOVE. The minimum size (in bytes) to consider a type for passing by reference instead of by value.
+    (pass_by_value_size_limit, "pass_by_value_size_limit": u64, 8),
```
A page size is 4kB, so this should be the maximum that this value should be. We can start with 256B, and then see what people set this to, if they modify it.

flip1995 reviewed (commented):
There's an empty `*.stdout` file that needs to be removed.

---

About `Vec` and `String`: You shouldn't get sizes bigger than 16/24Bytes on 64 bit machines, so they can always be safely passed by value. If someone sets the limit as low as that, there is another lint for `&Vec` and `&String` suggesting a slice or `&str`. So I don't think we need to deal with that here.

bors commented:
:umbrella: The latest upstream changes (presumably #6178) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

cauebs commented on clippy_lints/src/pass_by_ref_or_value.rs:
```diff
@@ -54,14 +54,41 @@ declare_clippy_lint! {
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for functions taking arguments by value, where
+    /// the argument type is not `Copy` (or an array) and large enough to be worth
+    /// considering passing by reference.
```
Do you mean "where the argument type is not **mutable**, not `Copy` and is large enough"? Otherwise I don't understand.

ebroto replied:
By "mutated" I meant "modified", but yeah, it can't be modified if it's not `mut` :) 

Note that we decided to only lint `Copy` types.

cauebs commented:
Ok, I think I addressed all your suggestions! Thanks a lot for the thorough feedback, @ebroto and @flip1995, I learned **a lot** more than I expected with this PR.

Let me know if there's anything else to improve here, but I might only work on it again next weekend.

ebroto reviewed (changes requested):
Only a small change and some tests to add, and this should be ready to merge!

cauebs commented on clippy_lints/src/pass_by_ref_or_value.rs:
```diff
@@ -0,0 +1,261 @@
+
+        let fn_body = cx.enclosing_body.map(|id| cx.tcx.hir().body(id));
+
+        let params: Box<dyn Iterator<Item = _>> = match fn_body {
+            Some(body) => box body.params.iter().map(Some),
+            None => box std::iter::repeat(None),
+        };
+
+        for ((input, &ty), param) in decl.inputs.iter().zip(fn_sig.inputs()).zip(params) {
```
Yeah, in Rust I instinctively avoid indexing, but I believe you're right, it can be more efficient and also clearer.

cauebs commented on tests/ui/large_types_passed_by_value.rs:
```diff
@@ -0,0 +1,36 @@
+    fn devoure_tuple(&self, tup: (LargeAndCopy, LargeAndCopy));
+    fn devoure_array_and_tuple_wow(&self, array: [u8; 6666], tup: (LargeAndCopy, LargeAndCopy));
+}
+
```
Out of curiosity: is `todo!()` any better than an empty block?

ebroto replied:
Oh not at all, I just have the habit of using it so I don't have to care about the return type, but in this case both work.

ebroto commented:
@bors r+

Thanks for your work on this lint!

bors commented:
:pushpin: Commit e8731a926c9a642ca1ddf5b52baf40e0a8873d53 has been approved by `ebroto`

<!-- @bors r=ebroto e8731a926c9a642ca1ddf5b52baf40e0a8873d53 -->
<!-- homu: {"type":"Approved","sha":"e8731a926c9a642ca1ddf5b52baf40e0a8873d53","approver":"ebroto"} -->

bors commented:
:hourglass: Testing commit e8731a926c9a642ca1ddf5b52baf40e0a8873d53 with merge 5c78d26e5e4d96d346e9c365b71ad59449782958...
<!-- homu: {"type":"BuildStarted","head_sha":"e8731a926c9a642ca1ddf5b52baf40e0a8873d53","merge_sha":"5c78d26e5e4d96d346e9c365b71ad59449782958"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/1303327713), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/1303326351), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1303358488)
Approved by: ebroto
Pushing 5c78d26e5e4d96d346e9c365b71ad59449782958 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ebroto","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/1303358488","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/1303327713","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/1303326351"},"merge_sha":"5c78d26e5e4d96d346e9c365b71ad59449782958"} -->

cauebs commented:
And thank you guys for the feedback and help!
